### PR TITLE
Grad boxen

### PR DESCRIPTION
**Pull request by chao-mu**

### CODE REVIEW

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+interface GraduationCardProps {
+    graduation: Graduation;
+}
+
+function getChartData(graduation: Graduation): UserSerie<Datum>[] {
+    const { ratingSystem: preferredSystem, startedAt } = graduation;
+
+    const ratingHistory =
+        graduation.ratingHistories?.[preferredSystem]
+            ?.filter(
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationCard.tsx:78]: Is this filter necessary? Aren't you already doing this on the backend?

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+        // There are potentials CORS issues with AWS
+        // https://github.com/bubkoo/html-to-image/issues/40
+        // https://stackoverflow.com/questions/42263223/how-do-i-handle-cors-with-html2canvas-and-aws-s3-images
+        // https://www.hacksoft.io/blog/handle-images-cors-error-in-chrome
+        domToPng(node, { backgroundColor })
+            .then((dataUrl) => {
+                setImageData(dataUrl);
+                /*
+                trackEvent(EventType.DownloadGradBox, {
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationCard.tsx:228]: Should this still be commented out?

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+                    previous_cohort: graduation.previousCohort,
+                    new_cohort: graduation.newCohort,
+                    dojo_score: graduation.score,
+                    graduated_at: graduation.createdAt,
+                });
+                */
+            })
+            .catch((error) => {
+                console.error('error :-(', error);
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationCard.tsx:237]: Can you change the error message to say `domToPng: ` instead?

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        link.click();
+        document.body.removeChild(link);
+    };
+
+    return (
+        <Stack>
+            {imageData ? (
+                <Stack>
+                    <img src={imageData} alt='Preview of graduation image' />
+                    <LoadingButton startIcon={<SaveAlt />} onClick={() => onDownload()}>
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationCard.tsx:263]: `onClick={onDownload}`

**Review by jackstenglein — CHANGES_REQUESTED** [review]

File: backend/user/graduate/main.go [hunk 5]
```diff
@@ -78,6 +84,28 @@ func Handler(ctx context.Context, event api.Request) (api.Response, error) {
...
+	ratingHistories := make(map[database.RatingSystem][]database.RatingHistory)
+	for rs, history := range user.RatingHistories {
+		for _, item := range history {
+			date, err := time.Parse(time.RFC3339, item.Date)
+			if err != nil {
+				return api.Failure(errors.Wrap(500, "Unable to parse rating histoy date", "", err)), nil
+			}
+
+			if date.After(startedAtTime) {
+				if date.After(createdAtTime) {
```
[the remark this answers is not in the document]
Comment (chao-mu) [reply backend/user/graduate/main.go:96]: My paranoid tendencies shining through.

File: frontend/src/calendar/displayDate.tsx [hunk 6]
```diff
@@ -53,7 +53,7 @@ export function getTimeZonedDate(
 
 export function toDojoDateString(
     date: Date,
-    timezone: string | undefined,
+    timezone?: string,
```
[the remark this answers is not in the document]
Comment (chao-mu) [reply frontend/src/calendar/displayDate.tsx:56]: Gladly.

File: frontend/src/profile/GraduationShareDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    onClose,
+    graduation,
+}: GraduationShareDialogProps) {
+    return (
+        <Dialog open={open} fullWidth>
+            <DialogTitle>Share your progress!</DialogTitle>
+            <DialogContent>
+                <Stack spacing={2}>
+                    <DialogContentText>
+                        Show off the hard work you've done and welcome the world to the
```
[the remark this answers is not in the document]
Comment (chao-mu) [reply frontend/src/profile/GraduationShareDialog.tsx:30]: MUCH better.

File: frontend/src/profile/GraduationShareDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    graduation: Graduation;
+}
+
+export default function GraduationShareDialog({
+    open,
+    onClose,
+    graduation,
+}: GraduationShareDialogProps) {
+    return (
+        <Dialog open={open} fullWidth>
```
[the remark this answers is not in the document]
Comment (chao-mu) [reply frontend/src/profile/GraduationShareDialog.tsx:25]: Good catch!

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+                width: '800px',
+                height: '500px',
+            }}
+        >
+            {children}
+        </Box>
+    );
+});
+
+interface GraduationCardProps {
```
[the remark this answers is not in the document]
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationCard.tsx:69]: Whoops... Thanks!

File: frontend/src/app/(scoreboard)/graduations/reports/[cohort]/[username]/page.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    params: {
+        cohort: string;
+        username: string;
+    };
+}
+
+export default async function Page({ params }: PageProps) {
+    const { username, cohort } = params;
+
+    const graduations = await listGraduationsByCohort(cohort);
```
[the remark this answers is not in the document]
Comment (chao-mu) [reply frontend/src/app/(scoreboard)/graduations/reports/[cohort]/[username]/page.tsx:16]: :+1:

File: frontend/src/components/graduations/GraduationCard.tsx (hunk 1 above)
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationCard.tsx:78]: In order to support rapid cohort advancement, I was going to grab the user's whole history if the graduation object for that cohort has none.

File: frontend/src/components/graduations/GraduationCard.tsx (hunk 2 above)
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationCard.tsx:228]: it should e moved and then uncommented.

File: frontend/src/components/graduations/GraduationCard.tsx (hunk 3 above)
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationCard.tsx:237]: Absolutely. Was a placeholder for proper error handling.

File: frontend/src/components/graduations/GraduationCard.tsx (hunk 4 above)
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationCard.tsx:263]: Force of paranoia.

File: frontend/src/App.tsx [hunk 11]
```diff
@@ -98,6 +99,7 @@ const router = createBrowserRouter(
                 <Route path='help' element={<HelpPage />} />
                 <Route path='tournaments'>
                     <Route index element={<TournamentsPage />} />
+                    <Route path='round-robin' element={<TournamentViewer />} />
```
Comment (jackstenglein) [inline frontend/src/App.tsx:102]: Why is this change here?

File: frontend/src/calendar/displayDate.tsx [hunk 12]
```diff
@@ -64,7 +64,7 @@ export function toDojoDateString(
 export function toDojoTimeString(
     date: Date,
     timezone: string | undefined,
-    timeFormat: TimeFormat | undefined,
+    timeFormat?: TimeFormat,
```
Comment (jackstenglein) [inline frontend/src/calendar/displayDate.tsx:67]: Can you revert this change as well? `timeFormat: TimeFormat | undefined` has different semantics than `timeFormat?: TimeFormat`. I want to ensure that if you call this function, you are passing in a time format, which may possibly be undefined. The latter allows you to not pass in a time format at all

File: frontend/src/components/graduations/GraduationLinkCardGrid.tsx [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+}: GraduationLinkCardGridProps) {
+    graduations.sort(
+        (a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime(),
+    );
+
+    return (
+        <Grid container p='16px' spacing={2} alignItems='center' justifyContent='center'>
+            {graduations.map((grad) => (
+                <Grid
+                    key={`${grad.username}//${grad.createdAt}`}
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationLinkCardGrid.tsx:22]: Bit of a nitpick, but could you remove the extra slash in the key?

File: frontend/src/navbar/NavbarMenu.tsx [hunk 14]
```diff
@@ -154,6 +154,11 @@ function allStartItems(toggleExpansion: (item: string) => void): NavbarItem[] {
                     icon: <MilitaryTech />,
                     href: '/tournaments/open-classical',
                 },
+                {
+                    name: 'Round Robin',
+                    icon: <MilitaryTech />,
+                    href: '/tournaments/round-robin',
```
Comment (jackstenglein) [inline frontend/src/navbar/NavbarMenu.tsx:160]: Why is this change included?

File: frontend/tmp.txt [hunk 15]
```diff
@@ -0,0 +1,9 @@
+/home/bestieboots/projects/dojo/old/grad/frontend/.env.test.local
```
Comment (jackstenglein) [inline frontend/tmp.txt:1]: What is this file for?

**Review by jackstenglein — CHANGES_REQUESTED** [review]

File: frontend/src/App.tsx (hunk 11 above)
Comment (jalpp) [reply frontend/src/App.tsx:102]: feel free to remove this in the grad branch, I'm already working on this in the tournament UI

File: frontend/src/navbar/NavbarMenu.tsx (hunk 14 above)
Comment (jalpp) [reply frontend/src/navbar/NavbarMenu.tsx:160]: feel free to remove this in the grad branch, I'm already working on this in the tournament UI

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+                            {' '}
+                            {displayName}
+                        </Typography>
+                    </Stack>
+                    <Stack direction='row' columnGap='1ch' flexWrap='wrap'>
+                        <Typography lineHeight={1} variant='h5'>
+                            on graduating to
+                        </Typography>
+                        <Typography lineHeight={1} variant='h5'>
+                            {newCohort}!
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationCard.tsx:226]: Why is this in a separate typography? This can be in the same one as `on graduating to` right?

File: frontend/src/profile/GraduationShareDialog.tsx [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+                            <ReportCanvas ref={reportRef}>
+                                <GraduationCard graduation={graduation} />
+                            </ReportCanvas>
+                        )}
+                    </Stack>
+                </Stack>
+            </DialogContent>
+            <DialogActions>
+                <Button onClick={() => onClose()}>Close</Button>
+                <LoadingButton loading={!imageData} onClick={() => onDownload(true)}>
```
Comment (jackstenglein) [inline frontend/src/profile/GraduationShareDialog.tsx:119]: I don't think the download button should close the dialog either. That surprised me when I was testing it earlier

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+                flexWrap='wrap'
+                justifyContent='center'
+                alignItems='center'
+                columnGap='2ch'
+                gridArea='header'
+            >
+                <Box>
+                    <Stack direction='row' columnGap='1ch' flexWrap='wrap'>
+                        <Typography lineHeight={1} variant='h5'>
+                            Congrats{' '}
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationCard.tsx:214]: Can we center this congrats text? It looks a bit weird left aligned

File: frontend/src/profile/GraduationShareDialog.tsx (hunk 17 above)
Comment (chao-mu) [reply frontend/src/profile/GraduationShareDialog.tsx]: Okay, good to know it surprised you - for sure that means we should change it.

Comment (chao-mu) [reply frontend/src/profile/GraduationShareDialog.tsx]: Ah. I see you already made the change.

File: frontend/src/components/graduations/GraduationCard.tsx (hunk 16 above)
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationCard.tsx:226]: This segmentation is to control where text chunks wrap.

File: frontend/tmp.txt (hunk 15 above)
Comment (chao-mu) [reply frontend/tmp.txt:1]: Shoot, that didn't mean to make it in. Ugh.

File: frontend/src/navbar/NavbarMenu.tsx (hunk 14 above)
Comment (chao-mu) [reply frontend/src/navbar/NavbarMenu.tsx:160]: The beta branch got mixed in when resolving merge conflicts.

File: frontend/src/calendar/displayDate.tsx (hunk 12 above)
Comment (chao-mu) [reply frontend/src/calendar/displayDate.tsx:67]: My mistake for missing it in the revert. Allowing not passing in whatever it was for the other change I reverted was indeed the intent, for it's usage in the link cards.

File: frontend/src/components/graduations/GraduationLinkCardGrid.tsx (hunk 13 above)
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationLinkCardGrid.tsx:22]: Done.

File: frontend/package.json [hunk 19]
```diff
@@ -36,9 +36,11 @@
         "chessground": "^9.1.1",
         "copy-to-clipboard": "^3.3.3",
         "date-fns": "^3.6.0",
+        "dom-to-image-more": "^3.3.1",
```
Comment (jackstenglein) [inline frontend/package.json:39]: Are we using this or just modern-screenshot?

File: frontend/src/components/graduations/GraduationCard.tsx [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+
+function StatLabel({ children, center }: { children: ReactNode; center?: boolean }) {
+    return (
+        <Typography
+            component='span'
+            fontSize='1rem'
+            variant='subtitle2'
+            color='text.secondary'
+            sx={{
+                textAlign: center ? 'center' : 'left',
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationCard.tsx:54]: Are there any StatLabels that aren't centered? It seems like this prop could be removed and it's just always centered

File: frontend/src/components/graduations/GraduationLinkCard.tsx [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+interface GraduationLinkCardProps {
+    graduation: Graduation;
+    to: string;
+}
+
+export const GraduationLinkCard = ({ graduation, to }: GraduationLinkCardProps) => {
+    const { newCohort, displayName, createdAt: graduatedAt } = graduation;
+
+    const dateStr = toDojoDateString(new Date(graduatedAt), undefined);
+    const timeStr = toDojoTimeString(new Date(graduatedAt), undefined, undefined);
```
Comment (jackstenglein) [inline frontend/src/components/graduations/GraduationLinkCard.tsx:19]: These dates/times should be passing in the viewer's timezone/timeformat

File: frontend/src/profile/GraduationShareDialog.tsx [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+    };
+
+    return (
+        <Dialog maxWidth='md' open={open} onClose={handleClose} fullWidth>
+            <DialogTitle>Share your progress!</DialogTitle>
+            <DialogContent>
+                <Stack spacing={2}>
+                    <DialogContentText>
+                        Show off your hard work and welcome the world to the Dojo!
+                        Download this image and share on social media.
```
Comment (jackstenglein) [inline frontend/src/profile/GraduationShareDialog.tsx:95]: Can we remove `Download this image and share on social media.`?

**Review by jackstenglein — CHANGES_REQUESTED** [review]

File: frontend/src/components/graduations/GraduationLinkCard.tsx (hunk 21 above)
Comment (chao-mu) [reply frontend/src/components/graduations/GraduationLinkCard.tsx:19]: This is not available while maintaining that it's publicly accessible pure SSR. I'd really like to maintain that design. Also, this isn't an exposed feature yet and is in there for debugging and future potential. I'd rather not fiddle with it more, but I would also like to keep it in.

File: frontend/src/profile/GraduationShareDialog.tsx (hunk 22 above)
Comment (chao-mu) [reply frontend/src/profile/GraduationShareDialog.tsx:95]: I would like to keep it in so there's a very specific call to action.

File: frontend/src/components/graduations/GraduationLinkCard.tsx (hunk 21 above)
Comment (jackstenglein) [reply frontend/src/components/graduations/GraduationLinkCard.tsx:19]: It has a 'use client' directive so isn't SSR right now anyway. But you can leave it as is for now

**Comment by chao-mu** [comment]
It requires some refactoring to remove poisonous imports that I didn't move in from the other branch if I recall correctly. Or it's there by mistake.








---- On Thu, 22 Aug 2024 17:32:48 -0400 ***@***.*** wrote ----



@jackstenglein commented on this pull request.

In frontend/src/components/graduations/GraduationLinkCard.tsx:

> +import { Graduation } from '@/database/graduation';
+import CohortIcon from '@/scoreboard/CohortIcon';
+import { Card, CardActionArea, CardContent, Stack, Typography } from ***@***.***/material';
+import Grid2 from ***@***.***/material/Unstable_Grid2';
+import Link from 'next/link';
+
+interface GraduationLinkCardProps {
+    graduation: Graduation;
+    to: string;
+}
+
+export const GraduationLinkCard = ({ graduation, to }: GraduationLinkCardProps) => {
+    const { newCohort, displayName, createdAt: graduatedAt } = graduation;
+
+    const dateStr = toDojoDateString(new Date(graduatedAt), undefined);
+    const timeStr = toDojoTimeString(new Date(graduatedAt), undefined, undefined);


It has a 'use client' directive so isn't SSR right now anyway. But you can leave it as is for now

**Review by jackstenglein — APPROVED** [review]
